### PR TITLE
SSO automation 

### PR DESCRIPTION
**Pull request by AniruddhaNayek**
Automation of Secondary Scheduler Operator

### CODE REVIEW

File: playbooks/roles/ocp-sso/tasks/main.yaml [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+- name: Setup custom catalog source
+  block:
+  - name: Include the global-secret-update role
+    include_role:
+        name: global-secret-update  
+
+  - name: Set fact variable for CatalogSource name
+    set_fact:
+      sso_catalogsource: "qe-app-registry"
```
Comment (SurajGudaji) [inline playbooks/roles/ocp-sso/tasks/main.yaml:16]: I think you can remove this task as the variable is already defined in the defaults/main.yaml and provide the value for "sso_catalogsource" as "qe-app-registry" instead of "redhat-operators" as defined in defaults/main.yaml.

File: playbooks/roles/ocp-sso/tasks/main.yaml [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+      name: make
+      state: present
+
+  - name: Include role for installation of Go lang
+    include_role:
+      name: golang-installation
+    vars:
+      go_tarball: "{{ sso_go_tarball }}"
+      golang_path: "/usr/local"
+
```
Comment (SurajGudaji) [inline playbooks/roles/ocp-sso/tasks/main.yaml:164]: you can add a precheck task to check the the sso_github_username and sso_github_token variable are not None.

File: playbooks/roles/ocp-sso/tasks/main.yaml [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+     when: terminating_namespace.stdout | int > 0
+     register: fix_terminating_namespace
+
+   - name: Wait for namespace to be deleted
+     shell: oc get ns openshift-secondary-scheduler-operator --no-headers | wc -l
+     register: verify_namespace
+     retries: 10
+     delay: 15
+     until: verify_namespace.stdout | int == 0
+
```
Comment (SurajGudaji) [inline playbooks/roles/ocp-sso/tasks/main.yaml:227]: For cleanup task can please refer this[ task](https://github.com/ocp-power-automation/ocp4-playbooks-extras/blob/main/playbooks/roles/ocp-coo/tasks/cleanup.yaml), and follow the same, because directly deleting the namespace is not right approach, delete all the resources which you have created during Deployment of SSO and then delete the namespace.

**Review by SurajGudaji — COMMENTED** [review]
Please add this role in the playbook/main.yml file.

File: playbooks/roles/ocp-sso/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+sso_catalogsource_image: ""
+sso_upgrade_channel: "stable"
+sso_schedulerImage: "registry.k8s.io/scheduler-plugins/kube-scheduler:v0.30.6"
+sso_go_tarball: "https://go.dev/dl/go1.22.4.linux-ppc64le.tar.gz"
+sso_e2e: false
+sso_e2e_github_repo: ""
+sso_e2e_github_branch: "master"
+sso_github_username: ""
+sso_github_token: ""
+sso_cleanup: true
```
[the remark this answers is not in the document]
Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/defaults/main.yml:15]: These variables "sso_github_username", "sso_github_token", "sso_github_repo" can be kept empty if not running the E2E tests for this operator . Also "sso_catalogsource_image" is only required when testing pre-release version of this operator or else it can be kept empty.

File: playbooks/roles/ocp-sso/tasks/main.yaml [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      sso_catalogsource: "qe-app-registry"
+
+  - name: Include role to create ImageContentSourcePolicy and CatalogSource
+    include_role:
+      name: set-custom-catalogsource
+    vars:
+      custom_catalogsource_name: "{{ sso_catalogsource }}"
+      custom_catalogsource_display_name: "Secondary Scheduler Operator"
+      custom_catalogsource_image: "{{ sso_catalogsource_image }}"
+  when: sso_catalogsource_image != '' and sso_catalogsource_image != None
```
[the remark this answers is not in the document]
Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/tasks/main.yaml:25]: The variable "sso_catalogsource_image" is only required if testing the pre-release version of this operator else the "sso_catalogsource" variable is considered which has a default value.

File: playbooks/roles/ocp-sso/tasks/main.yaml (hunk 1 above)
Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/tasks/main.yaml:16]: If validating a pre-release version for this operator the variable "sso_catalogsource" must be set as "qe-app-registry" as the openshift-test-private repo expects a catalogsource with this name or else if validating a GAed version the catalogsource as "redhat-operators" is considered

File: playbooks/roles/ocp-sso/tasks/main.yaml (hunk 2 above)
Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/tasks/main.yaml:164]: Added a task to ensure the above variables are not empty if running the E2E tests.

File: playbooks/roles/ocp-sso/tasks/main.yaml (hunk 3 above)
Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/tasks/main.yaml]: Updated the clean up task.

**Comment by AniruddhaNayek** [comment]
> Please add this role in the playbook/main.yml file.

Added.

File: playbooks/roles/ocp-sso/tasks/main.yaml (hunk 1 above)
Comment (SurajGudaji) [reply playbooks/roles/ocp-sso/tasks/main.yaml:16]: according to the playbook logic, the value provided in default as "redhat-operators" will never be used even if you want to validate GAed version

File: playbooks/roles/ocp-sso/defaults/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (SurajGudaji) [reply playbooks/roles/ocp-sso/defaults/main.yml:15]: please mark these variables as required if they don't have any default value

File: examples/all.yaml [hunk 6]
```diff
@@ -312,6 +312,20 @@ fio_git_token: ""
...
+sso_directory: "/tmp/sso"
+#sso_catalogsource_image: ""        # "brew.registry.redhat.io/rh-osbs/iib:327237"
+sso_upgrade_channel: "stable"
+sso_schedulerImage: "registry.k8s.io/scheduler-plugins/kube-scheduler:v0.30.6"
+sso_go_tarball: "https://go.dev/dl/go1.22.4.linux-ppc64le.tar.gz"
+sso_e2e: false
+sso_e2e_github_repo: "https://github.com/openshift/openshift-tests-private.git"
+sso_e2e_github_branch: "master"
+#sso_github_username: ""
+#sso_github_token: ""
```
Comment (varad-ahirwadkar) [inline examples/all.yaml:326]:
```suggestion
sso_github_username: ""
sso_github_token: ""
```

File: playbooks/roles/ocp-sso/tasks/main.yaml [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+      name: golang-installation
+    vars:
+      go_tarball: "{{ sso_go_tarball }}"
+      golang_path: "/usr/local"
+
+  - name: Precheck - Ensure sso_github_username and sso_github_token are not empty while running E2E
+    fail:
+      msg: "The variables 'sso_github_username' and 'sso_github_token' must be set when sso_e2e is true."
+    when: 
+      - sso_e2e | bool
```
Comment (varad-ahirwadkar) [inline playbooks/roles/ocp-sso/tasks/main.yaml:169]:
```suggestion
      
```
This block will get run only when `sso_e2e` is true

**Comment by varad-ahirwadkar** [comment]
Hi @AniruddhaNayek, can you please sign-off the commit ? Use your username and email id.

File: playbooks/roles/ocp-sso/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+sso_directory: "/tmp/sso"
+#sso_catalogsource_image: ""
+sso_upgrade_channel: "stable"
+sso_schedulerImage: "registry.k8s.io/scheduler-plugins/kube-scheduler:v0.30.6"
+sso_go_tarball: "https://go.dev/dl/go1.22.4.linux-ppc64le.tar.gz"
+sso_e2e: false
+sso_e2e_github_repo: "https://github.com/openshift/openshift-tests-private.git"
+sso_e2e_github_branch: "master"
+#sso_github_username: ""          
+#sso_github_token: ""
```
Comment (yussufsh) [inline playbooks/roles/ocp-sso/defaults/main.yml:14]: Why are some commented? Please remove if not required to be set and passed by the user. Also mark them as required in README if defaults are not set.

Comment (AniruddhaNayek) [reply playbooks/roles/ocp-sso/defaults/main.yml]: Updated the changes

File: playbooks/ocp-sso.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#ocp-sso.yml
+---
+- name: Installation of the Secondary Scheduler Operator and run e2e
+  hosts: bastion
+  tasks:
```
Comment (aman4433) [inline playbooks/ocp-sso.yml:5]: Can you please use this [format](https://github.com/ocp-power-automation/ocp4-playbooks-extras/blob/main/playbooks/ocp-coo.yml) to call the role

**Comment by aman4433** [comment]
Hey @AniruddhaNayek, Can you also mention this playbook in repo's `README.md`

File: playbooks/ocp-sso.yml (hunk 9 above)
Comment (AniruddhaNayek) [reply playbooks/ocp-sso.yml]: Configured the playbook with mentioned format
